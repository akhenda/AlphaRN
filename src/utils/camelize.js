/**
 * Camelize a string, cutting the string by multiple separators like
 * hyphens, underscores and spaces.
 * 
 * @param {text} string Text to camelize
 * @return string Camelized text
 */
export const camelize = (text) => {
  return text.replace(/^([A-Z])|[\s-_]+(\w)/g, (match, p1, p2) => {
    if (p2) return p2.toUpperCase();
    return p1.toLowerCase();        
  });
};

/**
 * Decamelizes a string with/without a custom separator (underscore by default).
 * 
 * @param str String in camelcase
 * @param separator Separator for the new decamelized string.
 */
export const decamelize = (str, separator) => {
  separator = typeof separator === 'undefined' ? '_' : separator;

  /* eslint-disable prefer-template */
  return str
    .replace(/([a-z\d])([A-Z])/g, '$1' + separator + '$2')
    .replace(/([A-Z]+)([A-Z][a-z\d]+)/g, '$1' + separator + '$2')
    .toLowerCase();
};
