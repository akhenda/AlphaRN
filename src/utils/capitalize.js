export const capitalizeWord = (word) => {
  if (!word) return;

  const letters = Array.from(word);
  return letters[0].toUpperCase() + letters.splice(1).join('').toLowerCase();
};

export const capitalizeWords = (sentence) => {
  const words = sentence.split(' ');

  return words.map(word => capitalizeWord(word.trim())).join(' ');
};
