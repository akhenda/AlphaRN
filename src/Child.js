import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { StyleSheet, View, Text, TouchableOpacity } from 'react-native';

import { capitalizeWords } from './utils/capitalize';

export default class Child extends Component {
  getCapitalizedText() {
    const { text } = this.props;
    if (text) return capitalizeWords(text);

    return 'You have not written anything!';
  }

  render() {
    return (
      <View testID='Child' style={styles.container}>
        <Text testID='ChildTitle' style={styles.title}>
          Capitalized Text:
        </Text>
        <Text testID='ChildOutput' style={styles.text}>
          {this.getCapitalizedText()}
        </Text>
        <TouchableOpacity testID='ClearText' onPress={this.props.onClear}>
          <Text style={styles.button}>
            CLEAR TEXT
          </Text>
        </TouchableOpacity>
      </View>
    );
  }
}

Child.propTypes = {
  text: PropTypes.string.isRequired,
  onClear: PropTypes.func.isRequired,
};

const styles = StyleSheet.create({
  container: {
    marginTop: 20,
    alignItems: 'center',
  },
  title: {
    fontSize: 18,
    marginBottom: 20,
  },
  text: {
    fontSize: 16,
    padding: 20,
  },
  button: {
    fontSize: 18,
    marginTop: 30,
    color: 'azure',
    borderRadius: 6,
    fontWeight: 'bold',
    paddingVertical: 10,
    paddingHorizontal: 15,
    backgroundColor: 'darkkhaki',
  },
});
