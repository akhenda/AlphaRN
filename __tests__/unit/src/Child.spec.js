import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Child from 'src/Child';
import * as Utils from 'src/utils/capitalize';


describe('Child', () => {
  let component;
  let onClearStub;
  const capitalizedString = 'mock string value';

  test('renders correctly', () => {
    const stringValue = 'random string';
    const snapshot = renderer.create(<Child text={stringValue} onClear={onClearStub} />).toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  beforeEach(() => {
    Utils.capitalizeWords = jest.fn();
    Utils.capitalizeWords.mockReturnValue(capitalizedString);
    onClearStub = jest.fn();
  });

  describe('when props.text is empty', () => {
    const stringValue = '';
    beforeEach(() => {
      component = shallow(<Child text={stringValue} onClear={onClearStub} />);
    });

    it('renders placeholder string', () => {
      const expectedString = 'You have not written anything!';
      expect(component.find('Text').getElements()[1].props.children).toEqual(expectedString);
    });

    it('does not call capitalizeWords', () => {
      expect(Utils.capitalizeWords.mock.calls.length).toBe(0);
    });

    describe('when clear text button is pressed', () => {
      it('should call onClear callback', () => {
        component.find('TouchableOpacity').simulate('press');
        expect(onClearStub).toBeCalled();
      });
    });
  });

  describe('when props.text is not empty', () => {
    const stringValue = 'random string';
    beforeEach(() => {
      component = shallow(<Child text={stringValue} onClear={onClearStub} />);
    });

    it('renders string with capitalizeWords', () => {
      expect(Utils.capitalizeWords).toBeCalledWith(stringValue);
      expect(component.find('Text').getElements()[1].props.children).toEqual(capitalizedString);
    });
  });
});
