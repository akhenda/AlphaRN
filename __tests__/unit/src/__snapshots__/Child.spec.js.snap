// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Child renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "marginTop": 20,
    }
  }
  testID="Child"
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 18,
        "marginBottom": 20,
      }
    }
    testID="ChildTitle"
  >
    Capitalized Text:
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 16,
        "padding": 20,
      }
    }
    testID="ChildOutput"
  >
    mock string value
  </Text>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    collapsable={undefined}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="ClearText"
    tvParallaxProperties={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "backgroundColor": "darkkhaki",
          "borderRadius": 6,
          "color": "azure",
          "fontSize": 18,
          "fontWeight": "bold",
          "marginTop": 30,
          "paddingHorizontal": 15,
          "paddingVertical": 10,
        }
      }
    >
      CLEAR TEXT
    </Text>
  </View>
</View>
`;
